homeassistant/components/statistics/sensor.py

```diff
@@ -433,24 +433,31 @@ async def async_added_to_hass(self) -> None:
             async_at_start(self.hass, self._async_stats_sensor_startup)
         )
 
-    def _add_state_to_queue(self, new_state: State) -> None:
-        """Add the state to the queue."""
-        self._available = new_state.state != STATE_UNAVAILABLE
-        if new_state.state == STATE_UNAVAILABLE:
-            self.attributes[STAT_SOURCE_VALUE_VALID] = None
-            return
-        if new_state.state in (STATE_UNKNOWN, None, ""):
-            self.attributes[STAT_SOURCE_VALUE_VALID] = False
-            return
-
+    def _add_state_to_queue(self, new_state: State, loaded_from_db: bool) -> None:
+        """Add the state to the queue (historic values loaded from the database are added on the opposite end)."""
```

Comment:
Lines should not be longer than 88 characters.

Reply:
Writing test cases to test for race conditions is at the very least challenging. Since the problem can't be reproduced deterministically, it is very hard to write code to reproduce the issue.

However, the problem appears with a comparably high likelihood (I would say surely more than 50%). I have tested the code with a test sensor / template as described in #67627. Restating the system a few times has consistently created the spikes. With the proposed code changes they are gone.

Furthermore, the code is running in my system since weeks without every producing another spike (as a custom component).

Writing a completely new test scenario to test for this particular race condition would create a lot of additional code. Currently, I'm trying to make the proposed changes smaller, because this was requested. Adding a large amount of test code would mean going in the opposite direction again.

If you would want to test for all possible timing-related issues that may still be there, this would require a very high amount of test code. The code does not contain anything that  is actively managing timing-based issues (at least I didn't find anything), which means there may be some more issues. However, those issues are probably a lot less likely to appear than the current one, where many people have reported them.

Finally, all existing tests are green with the proposed code changes. I would hope we can just get this integrated without insisting on corresponding tests.

Reply:
> Currently, I'm trying to make the proposed changes smaller, because this was requested. Adding a large amount of test code would mean going in the opposite direction again.

Yes that is very awesome, thanks for that.

> Adding a large amount of test code would mean going in the opposite direction again.

But keep in mind that test code is different from solving 3 issues in 1 PR. The point of adding test code is making sure we keep track of use cases and make sure we don't break them in the future. Since `statistics` is quite an important part of HA, and it really sucks to get mistakes in that integration (I mean that's why we are here now) so I think making sure we avoid these mistakes in the future will help.

We have a lot of possibilities to replicate real life behaviour like previous database values and setting states and freezing time, so I think it should not be impossible to add them.

I am not an expert in statistics, but I am happy to think along with you on how we can test this properly to avoid future regressions.

Reply:
I was able to adapt one of the existing tests to reproduce the error scenario, so we even have a test now.